File: Erlang/2_valid_parentheses.erl
isValid(S) ->
    isValid(S, []).

isValid([], []) -> true;
isValid([], _) -> false;
isValid([C|T], Stack) when C =:= $({; C =:= $[; C =:= ${ ->
    isValid(T, [C|Stack]);
isValid([C|T], [H|Stack]) ->
    case {C, H} of
        ($), $() -> isValid(T, Stack);
        ($], $[) -> isValid(T, Stack);
        ($}, ${) -> isValid(T, Stack);
        _ -> false
    end;
isValid(_, _) -> false.

File: Typescript/2_valid_parentheses.ts
interface IBracketMapping {
    [key: string]: string;
}

export function isValid(s: string): boolean {
    const stack: string[] = [];
    const mapping: IBracketMapping = {
        ')': '(',
        '}': '{',
        ']': '['
    };

    for (const char of s) {
        if (char in mapping) {
            const topElement = stack.pop();
            if (topElement !== mapping[char]) {
                return false;
            }
        } else {
            stack.push(char);
        }
    }

    return stack.length === 0;
}